Source/Error.swift

```diff
@@ -24,41 +24,540 @@
 
 import Foundation
 
-/// The domain used for creating all Alamofire errors.
-public let ErrorDomain = "org.alamofire.error"
-
-/// The custom error codes generated by Alamofire.
-public enum ErrorCode: Int {
-    case inputStreamReadFailed           = -6000
-    case outputStreamWriteFailed         = -6001
-    case contentTypeValidationFailed     = -6002
-    case statusCodeValidationFailed      = -6003
-    case dataSerializationFailed         = -6004
-    case stringSerializationFailed       = -6005
-    case jsonSerializationFailed         = -6006
-    case propertyListSerializationFailed = -6007
+/// `AFError` is the error type returned by the Alamofire framework. It encompasses a few
+/// different types of errors, each with their own associated reasons.
+///
+/// `.multipartEncodingFailed` errors are returned when some step in the multipart encoding process fails.
+///
+/// `.responseValidationFailed` errors are returned when a `validate()` call fails.
+///
+/// `.responseSerializationFailed` errors are returned when a response serializer encounters an error in the serialization process.
+public enum AFError: Error {
+
+    public enum MultipartEncodingFailureReason {
+        case bodyPartURLInvalid(at: URL)
+        case bodyPartFilenameInvalid(in: URL)
+        case bodyPartFileNotReachable(at: URL)
+        case bodyPartFileNotReachableWithError(url: URL, error: Error)
+        case bodyPartFileIsDirectory(at: URL)
+        case bodyPartFileSizeNotAvailable(at: URL)
+        case bodyPartFileSizeQueryFailedWithError(url: URL, error: Error)
+        case bodyPartInputStreamCreationFailed(for: URL)
+
+        case outputStreamCreationFailed(for: URL)
+        case outputStreamFileAlreadyExists(at: URL)
+        case outputStreamURLInvalid(url: URL)
+        case outputStreamWriteFailed(error: Error)
+
+        case inputStreamReadFailed(error: Error)
```

Comment:
Unfortunately enum cases don't offer any flexibility with associated value capture labels, otherwise some of those incongruous names would be elided and the case name would handle it. e.g. for `bodyPartURLInvalid` or the `*FailedWithError` cases, I wouldn't have a label at all, but I have to. I tried to make the cases read as grammatical phrases as much as possible while still maintaining uniqueness, but it wasn't possible in all cases. Plus there are some additional cases I had to handle that weren't in your previous rename. I added the `*FailedWithError` cases because you can't have duplicate case names that capture different values. Otherwise, like `outputStreamWriteFailed`, they wouldn't have that additional text. Unfortunately the Swift 3 naming guidelines don't offer any specific guidance for enum case names.